url8d3c4d17f5_catseye_JaC64_tgz-pJ8-JaC64J8/src/com/dreamfabric/jac64/SIDVoice6581.java

```diff
@@ -0,0 +1,740 @@
+  // last sample
+  public int lastSample = 0;
+  // ADSR is 8 bits
+  public int adsrVol = 0;
+
+  long nanos;
+  long total;
+
+  int pwidArr[] = new int[SIDVoice6581.GENLEN];
+
+  // This is a crappy method... should not be used later...
+  public int[] generateSound(long cycles) {
+    updateSound(cycles);
+    // Cheat down generated to 0
+    generated = 0;
+    return outBuffer;
+  }
+
+  public void updatePulseWidth(long cycles) {
+    if ((wave & PULSE) != 0 && lastCycles > 0) {
+      int pos = (int) ((cycles - lastCycles) /
+		       (SIDVoice6581.CYCLES_PER_SAMPLE + 10));
+      pwidArr[pos % SIDVoice6581.GENLEN] =
+	((((memory[sidbase + 3]& 0xf) << 8) +
+	  memory[sidbase + 2]) * waveLen) / 4095;
+    } else {
+      pwid = ((((memory[sidbase + 3]& 0xf) << 8) +
+	       memory[sidbase + 2]) * waveLen) / 4095;
+    }
+  }
+
+  public void updateSound(long cycles) {
+//     if (getSIDNo() == 0) {
+//       total = System.nanoTime() - nanos;
+//     }
+//     nanos = System.nanoTime();
+
+    int[] lookup;
+    int[] lookupW;
+    lastCycles = cycles;
+
+    // Always generate data - if cycle driven?!
+    // Maybe make the frq to be << 4 so that it is even more exact than
+    // now???
+    frq = (int) (0.5 + (1 << FRQ_BITS) * ((memory[sidbase + 1] << 8) +
+					  memory[sidbase]) * FRQCONV);
+    trueFrq = (1 << FRQ_BITS) *
+      ((memory[sidbase + 1] << 8) + memory[sidbase]) * FRQCONV;
+
+    pwid = ((((memory[sidbase + 3]& 0xf) << 8) +
+	     memory[sidbase + 2])
+	    * waveLen) / 4095;
+
+    if (frqFac != 16) {
+      frq = (frq * frqFac) >> 4;
+      trueFrq = (trueFrq * frqFac) / 16;
+    }
+
+    if (next != null) {
+      next_frq = next.frq;
+      // 	next_nextSample = next.nextSample;
+    }
+
+    int bIndex = generated;
+
+    switch(wave) {
+    case NONE:
+      for (int i = 0; i < GENLEN; i++) {
+	intBuffer[bIndex++] = 0xfff;
+	nextSample = (nextSample + frq) % WAVELEN;
+      }
+      break;
+    case TRIANGLE:
+      if (ring) {
+	boolean msb0 = false;
+	for (int i = 0; i < GENLEN; i++) {
+	  boolean msb = (msb0 = nextSample >= WAVELEN_HALF) ^
+	    (next_nextSample >= WAVELEN_HALF);
+
+	  long samp = nextSample;
+	  // Differ => modify position!
+	  if (msb0 != msb) {
+	    samp += msb0 ? -WAVELEN_HALF : WAVELEN_HALF;
+	  }
+	  intBuffer[bIndex++] = triangleWave[(int)(samp >> FRQ_BITS)];
+	  nextSample = (nextSample + frq) % WAVELEN;
+	  next_nextSample = (next_nextSample + next_frq) % WAVELEN;
+	}
+      } else if (!sync) {
+	for (int i = 0; i < GENLEN; i++) {
+	  intBuffer[bIndex++] = triangleWave[(int)(nextSample >> FRQ_BITS)];
+	  nextSample = (nextSample + frq) % WAVELEN;
+	}
+      } else {
+	// SYNCH
+	for (int i = 0; i < GENLEN; i++) {
+	  intBuffer[bIndex++] = triangleWave[(int)(nextSample >> FRQ_BITS)];
+	  nextSample = (nextSample + frq) % WAVELEN;
+	  next_nextSample += next_frq;
+	  if (next_nextSample > WAVELEN) {
+	    nextSample = 0;
+	    next_nextSample -= WAVELEN;
+	  }
+	}
+      }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 493:509, 512:528.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.